src/react/reconcilation.js

```diff
@@ -794,21 +794,35 @@ export class Reconciler {
     let length = value.args.length;
     if (length > 0) {
       let newBranchState = new BranchState();
+      let args = [];
+      let shouldCreateNewAbstract = false;
       // TODO investigate what other kinds than "conditional" might be safe to deeply resolve
       for (let i = 0; i < length; i++) {
-        value.args[i] = this._resolveDeeply(
+        let arg = this._resolveDeeply(
           componentType,
           value.args[i],
           context,
           "NEW_BRANCH",
           newBranchState,
           evaluatedNode
         );
+        args.push(arg);
+        if (value.args[i] !== arg) {
+          shouldCreateNewAbstract = true;
+        }
       }
       let didBranch = newBranchState.applyBranchedLogic(this.realm, this.reactSerializerState);
       if (didBranch && branchState !== null) {
         branchState.mergeBranchedLogic(newBranchState);
       }
+      if (shouldCreateNewAbstract) {
+        invariant(
+          value.kind === "conditional",
+          "expected conditional abstract value when resolving value in React reconcilation"
+        );
+        let [condition, left, right] = args;
```

Comment:
You can assert that condition is an abstract value here because value.args[0] is guaranteed to be abstract and _resolveDeeply returns an abstract when called on an abstract. Please mention all this in the comment explaining why the assertion is true.